### PR TITLE
Bugfix/prtide

### LINKED ISSUE #225: Regtest tp2.18: ww3_prtide_current screen output not reproducible with MPI
After the recent updates to add tidal input to regtest `ww3_tp2.18` (https://github.com/NOAA-EMC/WW3/commit/7620443e9c2d6401c49cdc97dcbad3ed78550f46), the screen output from the `ww3_prtide_current` program is not reproducible when running in MPI mode.

The output includes lines titled "BAD POINT" which are written out by individual processors in MPI mode. When running ww3_prtide as an MPI program, the order of this output is arbitrary which shows up as a difference in the `ww3_prtide_current.out` files when doing a regtest comparison.

Performing a "sort" on the output files helps a bit, but not all output gets written to a separate newline, so this does not completely resolve the differences without manual editing.

### PR DESCRIPTION
**Pull request by mickaelaccensi**
# Pull Request Summary
the screen output from the ww3_prtide_current program is not reproducible when running in MPI mode

## Description
set mpi gather function to only output messages from processor NAPOUT instead of each processor

### Issue(s) addressed
* Is there an issue associated with this development (bug fix, enhancement, new feature)? 
- fixes #225 
- fixes also the issue #402 
- fixes also the issue #407 

### Check list  
* Is your feature branch up to date with the authoritative repository (NOAA/develop)? YES
* Make sure you have checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop), [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop) and [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) YES
* Reviewers: @ukmo-ccbunney @aliabdolali 

### Testing
* How were these changes tested? regtest ww3_tp2.18
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) 
* If a new feature was added, was a new regression test added?
* Have regression tests been run? only ww3_tp2.18
* Which compiler / HPC you used to run the regression tests in the PR? intel 2018
* Please provide the summary output of matrix.comp (_matrix.Diff.out_, _matrixCompFull.out_ and _matrixCompSummary.out_):    

[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6577571/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6577572/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6577573/matrixDiff.txt)

### CODE REVIEW

File: model/ftn/ww3_prnc.ftn [hunk 1]
```diff
@@ -1823,8 +1854,8 @@
 !
                 IF (NFCOMP.EQ.2) THEN
 !/O3             IF ( IAPROC .EQ. NAPOUT )    WRITE (NDSO,976) ' (2) '
-                    CALL INTERP(YC, JX21, JX22, JY21, JY22, XD11, XD12,&
-                                             XD21, XD22, FILLVALUE, FA)
+                    CALL INTERP(MXM, MYM, YC, JX21, JX22, JY21, JY22,  &
```
Comment (ukmo-ccbunney) [inline model/ftn/ww3_prnc.ftn:1860]: Is this a separate bug fix?
It looks like the call to interp was missing the dimension variables MXM and MYM.

**Review by ukmo-ccbunney — COMMENTED** [review]
The changes to `ww3_prnc.ftn` look sensible to me Mickael. I just had one comment regarding a call to INTERP that has changed.

It looks like the only changes to `ww3_prtide.ftn` are splitting long lines and changing some format specifiers. Is that correct?

I',m just running some tests at my end.

File: model/ftn/ww3_prnc.ftn (hunk 1 above)
Comment (mickaelaccensi) [reply model/ftn/ww3_prnc.ftn:1860]: yes, it was crashing when compiling so I've corrected it

File: model/ftn/ww3_prtide.ftn [hunk 2]
```diff
@@ -501,14 +505,15 @@
...
+                  WRITE(NDSO,'(A,I8,F10.2,A,F10.2)') 'BAD: ', INDMAX(I), MAXVALCON(I), &
+                        '/', ABS(TIDAL_CONST(IX,IY,INDMAX(I),1:2,1))
                 END IF
                 BADPOINTS(IX,IY) = BADPOINTS(IX,IY) + (1-TIDEOK)
               END DO
 
               IF (BADPOINTS(IX,IY).GT.0) THEN 
                 NBAD = NBAD +1
-                WRITE(NDSE,*) 'BAD POINT:',IX,IY,NBAD, &
+                WRITE(NDSE,'(A,I8,I8,F10.2,A,F10.2)') 'BAD POINT:',IX,IY,NBAD, &
```
Comment (ukmo-ccbunney) [inline model/ftn/ww3_prtide.ftn]: The format specifier is wrong here; I think you need another I8 (for NBAD), i.e.
``` '(A,I8,I8,I8,F10.2,A,F10.2)') ```

Comment (ukmo-ccbunney) [reply model/ftn/ww3_prtide.ftn]: Furthermore, the slice `TIDAL_CONST(IX,IY,:,1,1)` is multiple values, not a single float, and this is causing runtime errors as there are not enough format specifiers. An example line (produced by using list directed I/O) looks like this:

```
BAD POINT:           2           1           1  -327.669983       2.33866073E-12   2.27044360E-12   2.26722764E-12   2.25777578E-12   2.49276901E-12   2.32712092E-12   2.32340861E-12   2.39584112E-12   2.06419994E-12   2.35532514E-12   2.40584549E-12   2.05155250E-12   2.04318875E-12   2.16234626E-12   2.15818834E-12   2.20166940E-12   1.93449011E-12   2.05661768E-12   2.03594628E-12 ##  -327.669983       2.33866073E-12   2.27044360E-12   2.26722764E-12   2.25777578E-12   2.49276901E-12   2.32712092E-12   2.32340861E-12   2.39584112E-12   2.06419994E-12   2.35532514E-12   2.40584549E-12   2.05155250E-12   2.04318875E-12   2.16234626E-12   2.15818834E-12   2.20166940E-12   1.93449011E-12   2.05661768E-12   2.03594628E-12
BAD:        1      1.00/    327.67```

**Review by ukmo-ccbunney — CHANGES_REQUESTED** [review]
@mickaelaccensi - I ma having some problems with the "BAD" output line due to the new format specifiers. See specific review comments.

File: model/ftn/ww3_prtide.ftn (hunk 2 above)
Comment (mickaelaccensi) [reply model/ftn/ww3_prtide.ftn]: I've pushed the corrected version but I don't have any BAD POINT statements in the log files.. so I'm not able to confirm it's ok now. Do you have this output with regtest ww3_tp2.18 ?

Comment (ukmo-ccbunney) [reply model/ftn/ww3_prtide.ftn]: @mickaelaccensi I get lots of "BAD POINT" output when I run that test. It's odd that you don't get any.
Also, the order of the "BAD POINT" output is still not reproducible between two identical runs.

I have attached a zip file here with the output from two runs of ww3_tp2.18/TIDE_MPI: [ww3_prtide_current_out.zip](https://github.com/NOAA-EMC/WW3/files/6584486/ww3_prtide_current_out.zip)

Comment (ukmo-ccbunney) [reply model/ftn/ww3_prtide.ftn]: I notice that there are a lot of values of 327.67 or -327.67 in those "BAD" points that I sent you - they look suspiciously like the netCDF `fille_value` of -32678 (2^15) scaled by 0.01... Are there perhaps some invalid points being read from the netCDF?

Comment (JessicaMeixner-NOAA) [reply model/ftn/ww3_prtide.ftn]: I don't get any "BAD POINT" in the ww3_tp2.18 ww3_prtide_current.out either.  I did run the two tests twice and confirmed they reproduce each other now for NCEP.

File: model/ftn/ww3_prnc.ftn (hunk 1 above)
Comment (aliabdolali) [reply model/ftn/ww3_prnc.ftn:1860]: Is there an issue associated with this fix?

**Comment by aliabdolali** [comment]
I ran the regtests on our hpc, all ran successfully with the expected non-b4b tests.
[PRTIDE.zip](https://github.com/NOAA-EMC/WW3/files/6609745/PRTIDE.zip)

```
mww3_test_03
mww3_test_07
ww3_tp2.10
ww3_tp2.15
ww3_tp2.18
```

@ukmo-ccbunney Do you still have BAD POINTS at your end?

**Comment by JessicaMeixner-NOAA** [comment]
Moving forward tp2.18 should be off the NCEP list as a "normal" not b4b! Thanks for that @mickaelaccensi

File: model/ftn/ww3_prnc.ftn (hunk 1 above)
Comment (mickaelaccensi) [reply model/ftn/ww3_prnc.ftn:1860]: I've just created an issue #402 to document it

**Comment by ukmo-ccbunney** [comment]
@mickaelaccensi I am still getting lots of BAD POINT messages, as per my message above. I have a suspicion that it is due to the `Fill_Value` being read in from the input netCDF file. 

I am getting this with both the Cray and the GNU compiler.

```
BAD:        1      1.00/    327.67
�գC
 BAD POINT: 58,  1,  57,  -327.669983,  4.687815755E-14,  6.086700832E-14,  8.840115597E-14,  4.639287205E-14,  9.353529371E-14,  5.991358803E-14,  7.043603439E-14,  9.749133738E-14,  8.284382024E-14,  6.117295662E-14,  7.156954743E-14,  5.715787137E-14,  6.435384994E-14,  5.974144383E-14,  6.371199548E-14,  7.099765112E-14,  6.359484743E-14,  6.822032495E-14,  6.024069861E-14 ## -327.669983,  4.687815755E-14,  6.086700832E-14,  8.840115597E-14,  4.639287205E-14,  9.353529371E-14,  5.991358803E-14,  7.043603439E-14,  9.749133738E-14,  8.284382024E-14,  6.117295662E-14,  7.156954743E-14,  5.715787137E-14,  6.435384994E-14,  5.974144383E-14,  6.371199548E-14,  7.099765112E-14,  6.359484743E-14,  6.822032495E-14,  6.024069861E-14
```

Also, the ordering of the "BAD POINTS" is still not comparable between two consecutive runs.

**Comment by aliabdolali** [comment]
@ukmo-ccbunney Can you test tp2.18 again and see if the problem is solved?

**Comment by mickaelaccensi** [comment]
I found the problem raised by Chris, I've fixed it and I'm running the regtests now. I'll create a new issue to link it in the PR

**Comment by mickaelaccensi** [comment]
branch updated with last develop branch version and with bug correction #407 raised by Chris.
@ukmo-ccbunney  and @aliabdolali , can you review it again ?

here are the regtests comparison files
-for intel compiler : 
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6632558/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6632559/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6632560/matrixDiff.txt)

-for gnu compiler : 
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6632586/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6632587/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6632588/matrixDiff.txt)

you can see that the output of develop branch with gnu compiler contains the "BAD POINTS" statements and it's corrected in the prtide branch outputs

**Comment by aliabdolali** [comment]
@mickaelaccensi thanks for fixing the bad point issue. I checked it at our end it looks ok.
@ukmo-ccbunney could you check ww3_tp2.18 tests and finalize your review, so we can move on and merge this PR?

**Comment by ukmo-ccbunney** [comment]
@mickaelaccensi - Apologies, but I am still having some problems with this. I am getting runtime **floating point exception** errors when running with debug flags. It looks like there are some NaN values occurring in the `SETVUF_FAST` subroutine of w3tidemd.

I put this debug line at line 636 of w3tidemd:

```fortran
DO J=TIDE_INDEXJ(K),TIDE_INDEXJ(K)+NJ(K)-1
          L2=TIDE_INDEXJK(J)

          PRINT*,'CB :',J,L2,FA(K),FA(L2),ABS(COEF_CON(J))

          FA(K)=FA(K)*FA(L2)**ABS(COEF_CON(J))
          vA(k)=vA(K)+COEF_CON(J)*vA(L2)
          UA(K)=UA(k)+COEF_CON(J)*UA(L2)
          END DO   ! J
        END DO     ! K
```

and I see `NaN` values appearing in some of the `FA` entries
```
 Time : 2019/02/01 00:00:00 UTC
 Number of remaining bad points:           0
 CB :           1          19   1.00000000                  NaN   2.00000000    
 CB :           2          13              NaN  -7.94284777E-13   1.00000000    
 CB :           3          41   1.00000000      0.999180675       1.00000000    

```

I'll continue investigating and see if I can find out where they are coming from.

**Comment by ukmo-ccbunney** [comment]
OK - I think I know why the NaNs are appearing: The `FA` array is not initialised in the `SETVUF_FAST` routine and has only some of it's elements updated in the loop starting at line 588 (depending on the condition ` IF (TIDE_INDEX2(L).EQ.K)`)

If I add `FA(:) = 0` at the beginning of the routine, the calculations are all ok.

I am going to run the regression test again to confirm.

**Comment by ukmo-ccbunney** [comment]
`UA` and `VA` probably need initialising too.

**Comment by ukmo-ccbunney** [comment]
OK - after initialising FA, UV and VA I am now getting B4B results between runs and no floating point exception errors.

**Comment by ukmo-ccbunney** [comment]
Finally  (and not directly related to this PR, but you might want to fix it here) - I noticed this line is causing a runtime error due to the format specification at `1003` not specifying an integer. I guess in this case could just remove IERR for L763?

https://github.com/NOAA-EMC/WW3/blob/4a268f2948ec94d034a84537e0f9ee622f8d9917/model/ftn/ww3_prtide.ftn#L763

https://github.com/NOAA-EMC/WW3/blob/4a268f2948ec94d034a84537e0f9ee622f8d9917/model/ftn/ww3_prtide.ftn#L794-L796

**Comment by aliabdolali** [comment]
@ukmo-ccbunney thanks for the tests and pointing to the problem. @mickaelaccensi could you modify the code, as Chris suggested? 
Thanks, both for helping.

**Comment by mickaelaccensi** [comment]
corrections added. @aliabdolali you can review and (squash) merge now

**Comment by aliabdolali** [comment]
@ukmo-ccbunney can you test ww3_tp2.18 and provide your final review? It ran finre on our hpc.

**Review by ukmo-ccbunney — APPROVED** [review]
@mickaelaccensi - the updates all ran OK on our HPC and the results are now bit-for-bit (with no "BAD POINT" warnings). Nice one! :)

**Comment by aliabdolali** [comment]
> @mickaelaccensi - the updates all ran OK on our HPC and the results are now bit-for-bit (with no "BAD POINT" warnings). Nice one! :)

Awesome @mickaelaccensi and @ukmo-ccbunney.

**Review by aliabdolali — APPROVED** [review]

**Comment by aliabdolali** [comment]
I checked ww3_tp2.18 again and confirm that it is now b4b. so I'll remove this test from the non-b4b list.

**Pull request merged**